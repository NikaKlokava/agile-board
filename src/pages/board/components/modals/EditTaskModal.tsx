import { FieldArray, Form, Formik } from "formik";
import { Button } from "../../../../shared/components/button";
import { DescriptionField } from "../../../../shared/components/description/DescriptionField";
import { Input } from "../../../../shared/components/input";
import { ModalWrapper } from "../../../../shared/components/modal_wrapper";
import { Select } from "../../../../shared/components/select";
import { FieldName } from "../../../../shared/components/field_name";
import { EditTaskSchema } from "../../../../utils/utils";
import { AddBtn } from "../../../../shared/components/add_button";
import { editTask, moveTask } from "../../../../redux/reducers/tasksSlice";
import { useAppDispatch, useAppSelector } from "../../../../redux/hooks/hook";
import { v4 as uuidv4 } from "uuid";
import cl from "./modal_styles.module.css";
import { updateTaskData } from "../../../../redux/reducers/tasksSlice";
import { CloseIcon } from "../../../../shared/components/close_icon/CloseIcon";

type Props = {
  taskUuid: string;
  onClose: () => void;
  onTaskModalClose: () => void;
};

export const EditTaskModal = ({
  taskUuid,
  onClose,
  onTaskModalClose,
}: Props) => {
  const activeBoard = useAppSelector((state) => state.activeBoard);

  const tasks = useAppSelector((state) => state.tasks.tasks);

  const taskIndex = tasks.findIndex((task) => task.uuid === taskUuid);
  const task = tasks[taskIndex];

  const columnIndex = activeBoard.columns?.findIndex(
    (column) => column.uuid === task.columnUuid
  );
  const column = activeBoard.columns?.[columnIndex]?.title;

  const dispatch = useAppDispatch();

  const checkedSubtasks = task?.subtasks?.reduce(
    (accum: boolean[], current) => {
      return [...accum, current.checked];
    },
    []
  );

  const onSubmit = (values: EditTaskType) => {
    const columnIndex = activeBoard.columns?.findIndex(
      (column) => column.title === values.columnTitle
    );
    const columnUuid = activeBoard.columns?.[columnIndex].uuid;
    const time = new Date().getTime();
    const updatedTask = {
      ...task,
      columnUuid,
      title: values.title,
      description: values.description,
      subtasks: values.subtasks,
      time,
    };
    columnUuid && dispatch(moveTask({ taskUuid, columnUuid, time }));
    dispatch(editTask(updatedTask));
    dispatch(updateTaskData({ updatedTask }));

    onClose();
    onTaskModalClose();
  };

  const initialTaskData = {
    title: task?.title,
    description: task?.description,
    subtasks: task?.subtasks?.reduce((accum: string[], current) => {
      return [...accum, current.text];
    }, []),
    columnTitle: column,
  };

  return (
    <ModalWrapper onWrapperClick={onClose}>
      <Formik
        initialValues={initialTaskData}
        validationSchema={EditTaskSchema}
        onSubmit={(values) => {
          const subtasks = values.subtasks
            .filter((subtask) => subtask && subtask.trimStart().length !== 0)
            .map((subtask) => subtask.trimStart());

          const newSubtasks = subtasks.reduce(
            (accum: { text: string; checked: boolean }[], current, index) => {
              return [
                ...accum,
                {
                  text: current,
                  checked: checkedSubtasks?.[index]
                    ? checkedSubtasks[index]
                    : false,
                  uuid: uuidv4(),
                },
              ];
            },
            []
          );
          onSubmit({
            ...values,
            subtasks: newSubtasks,
          });
        }}
      >
        {(props) => (
          <Form
            onSubmit={(e) => {
              e.preventDefault();
              props.handleSubmit();
            }}
            className={cl.form_container}
          >
            <CloseIcon onClose={onClose} />
            <h2 className={cl.modal_title} data-testid="new-task-modal">
              Edit Task
            </h2>
            <FieldName formikName="title" fieldName={"Task name"} />
            {props.errors.title && props.touched.title && (
              <p style={{ color: "red" }}>{props.errors.title}</p>
            )}
            <DescriptionField />
            <FieldArray
              name="subtasks"
              render={(arrayHelpers) => (
                <>
                  <div className={cl.container}>
                    <p className={cl.title}>Subtasks</p>
                    {props.values.subtasks?.map((_, index) => (
                      <Input
                        key={index}
                        formikName={`subtasks[${index}]`}
                        remove={arrayHelpers.remove}
                        index={index}
                        checked={checkedSubtasks}
                      />
                    ))}
                    <AddBtn
                      add={arrayHelpers.push}
                      text="Add New Subtask"
                      testid={"add-new-subtask-btn"}
                    />
                  </div>
                </>
              )}
            />
            <Select />
            <Button
              text="Save Edit"
              withIcon={false}
              newClass="center"
              testid={"create-task-btn"}
              type="submit"
            />
          </Form>
        )}
      </Formik>
    </ModalWrapper>
  );
};
