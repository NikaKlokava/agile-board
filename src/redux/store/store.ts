import { legacy_createStore } from "redux";
import { boardsReducer } from "../reducers/boardReducer";

const store = legacy_createStore(boardsReducer);

export default store;

// export type RootState = ReturnType<typeof store.getState>

// export interface RootState {
//     user: UserState;
//     loader: LoaderState;
//   }
// export type AppDispatch = typeof store.dispatch
