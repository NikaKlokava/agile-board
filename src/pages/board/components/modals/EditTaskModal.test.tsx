import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import store from "../../../../redux/store/store";
import { EditTaskModal } from "./EditTaskModal";

describe("Test the EditTaskModal component", () => {
  test("The EditTaskModal renders correctly", () => {
    const editTaskModalSnap = renderer
      .create(
        <Provider store={store}>
          <EditTaskModal
            taskUuid={"test id"}
            onTaskModalClose={() => {
              console.log("test");
            }}
          />
        </Provider>
      )
      .toJSON();
    expect(editTaskModalSnap).toMatchSnapshot();
  });
});
