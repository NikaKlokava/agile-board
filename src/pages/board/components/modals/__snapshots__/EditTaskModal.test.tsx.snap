// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the EditTaskModal component The EditTaskModal renders correctly 1`] = `
Array [
  <div
    className="modal_wrapper"
  />,
  <div
    className="modal_container"
  >
     
    <h2
      className="modal_title"
      data-testid="new-task-modal"
    >
      Add New Task
    </h2>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Task name
      </p>
      <input
        autoComplete="off"
        name="title"
        onChange={[Function]}
        placeholder="e.g Take coffee break"
        spellCheck={false}
        type="text"
      />
    </div>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Description
      </p>
      <textarea
        className="textarea"
        name="description"
        onChange={[Function]}
        placeholder="e.g. It's always good to take a break..."
        spellCheck={false}
      />
    </div>
    <div
      className="container style_container"
    >
      <p
        className="title"
      >
        Subtasks
      </p>
      <div
        className="button"
        data-testid="add-new-subtask-btn"
        onClick={[Function]}
      >
        <div
          className="button_icon"
        />
        <div
          className="button_text"
        >
          Add New Subtask
        </div>
      </div>
    </div>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Current Status
      </p>
      <select
        name="columnTitle"
        onChange={[Function]}
      />
    </div>
    <div
      className="button center"
      data-testid="create-task-btn"
      onClick={[Function]}
    >
      <div
        className="button_text"
      >
        Save Edit
      </div>
    </div>
  </div>,
]
`;
